src/solvers/dgsem_tree/dg_3d_compressible_euler.jl

```diff
@@ -0,0 +1,264 @@
+
+# From here on, this file contains specializations of DG methods on the
+# TreeMesh3D to the compressible Euler equations.
+#
+# The specialized methods contain relatively verbose and ugly code in the sense
+# that we do not use the common "pointwise physics" interface of Trixi.jl.
+# However, this is currently (November 2021) necessary to get a significant
+# speed-up by using SIMD instructions via LoopVectorization.jl.
+#
+# TODO: SIMD. We could get even more speed-up if we did not need to permute
+#             array dimensions, i.e., if we changed the basic memory layout.
+#
+# We do not wrap this code in `@muladd begin ... end` block. Optimizations like
+# this are handled automatically by LoopVectorization.jl.
+
+
+# We specialize on `PtrArray` since these will be returned by `Trixi.wrap_array`
+# if LoopVectorization.jl can handle the array types. This ensures that `@turbo`
+# works efficiently here.
+@inline function split_form_kernel!(_du::PtrArray, u_cons::PtrArray,
+                                    element, mesh::TreeMesh{3},
+                                    nonconservative_terms::Val{false},
+                                    equations::CompressibleEulerEquations3D,
+                                    volume_flux::typeof(flux_shima_etal),
+                                    dg::DGSEM, cache,
+                                    alpha=true)
```

Comment:
This variant seems like the new, optimized version will always be used when using the split form kernel with the Shima et al. flux. Wouldn't it be nice - maybe at least for a transition period - do make this opt in? E.g. by providing a `Fast` wrapper type such as
```julia
struct Fast{T}
  entity::T
end
```
such that you could use `volume_flux=Fast(flux_shima_etal)` in the elixir?

I am feeling a little uneasy about having a special, super-optimized kernel being the default for just a single, selected flux, especially if new users want to experiment with it. If the use of the optimized version were explicit rather than implicit, it would imho reduce the chance of "suprises" (POLA).

Reply:
We can keep this in our minds and discuss it later. I would prefer to resolve the remaining comments here to allow me to proceed with other setups. We can revisit the decision of "slow by default" vs. "fast by default" later before merging the target branch `hr/SIMD` into `main`. Does that sound okay to you?

Reply:
> We can keep this in our minds and discuss it later. [...] We can revisit the decision of "slow by default" vs. "fast by default" later before merging the target branch hr/SIMD into main.

Sounds good to me! I see you arguments about making Trixi "fast by default", but here it clashes with "easy to understand by default". At the moment, I don't have a good suggestion how to reconcile these two conflicting goals for this particular case. My suggestion with the `Fast` wrapper was partly geared towards achieving this aim, by providing a "standardized" way to activate "invasive" performance optimization.

We had similar discussions about `wrap_array_native` vs. `wrap_array` and `multiply_pointwise!`. In those cases we opted for "fast by default", after some discussions. However, I consider this case to be slightly different since it involves an optimization that chips away some of the usual composability of Trixi.

Reply:
@trixi-framework/principal-developers What's your opinion on this?

Reply:
As a developer, I lean towards slow by default (for consistency and to avoid issues when benchmarking or comparing methods or solvers). As a user, I think I would prefer fast by default though. 

I think I lean towards "fast by default", so users don't need to learn and specify a "Fast" option.

Reply:
I guess I am not clear on the plan here with all the optimization capabilities. Would we make a super fast performance version of the fluxes to act as "default" for all the available equation systems (e.g. compressible Euler, GLM-MHD, shallow water)? Then if others want to look at the details they can pick other non-optimized variants? 

Or do we consider the compressible Euler equations to be the "flagship" and/or "benchmark" equations. So there we highlight how optimized it could be with the `shima et al` flux but leave everything else untouched with maybe some comments about possible optimizations?

TBH, I am getting confused with all these options about the combinations of fluxes and their respective optimizations.   I am worried that new users would be confused as well. What flux should I use? Why is this flux so much faster than that flux? There is already quite a hurdle convincing people to use the split formulation I worry that these extra questions / confusions will drive them away further.

Reply:
As you can see here, these optimizations are relatively involved due to Trixi's current structure and current limitations of LoopVectorization.jl. This PR implements stuff only for `flux_shima_etal` in 3D on the Cartesian mesh. I have similar stuff collected in `hr/SIMD` for 

- `flux_shima_etal` and `flux_ranocha`
- Cartesian and curved meshes (no `DGMulti`)
- 2D and 3D
- the compressible Euler equations

I would like to ship these as "fast by default" with Trixi.jl without any need for users to modify stuff.

Reply:
I am strongly for "slow by default". The code should be as accessible as possible - that is its main goal imo. 

We should think about making the "fast" option as convenient and simple to use as possible imo. But imo, it should not be the (first) version that is the default.

Reply:
If you want to keep it simple, do you want to get this into Trixi.jl at all?

Reply:
Yes! It is great work! We discussed this in the Trixi meeting, that it would be really perfect to have a "fast()" option, as @sloede mentioned above.

Reply:
I have discussed it with Hendrik yesterday. For now, we will introduce [`_turbo` versions for the two fluxes](https://github.com/trixi-framework/Trixi.jl/pull/991/files#diff-707db742ac66e01ef28a7a764728a1140e47cb0c684d4bece2a8839892402b6eR250-R273) where we have such optimizations available. This will allow us to immediately use them without having to change anything.

As @ranocha pointed out, having a `Fast(flux_shima_etal)` wrapper has the downside that people might start using it indiscriminately, not knowing when it will actually do something and when it won't. the `_turbo` version has similar issues, but it is less general and easier to deprecate should we at some point decide to use another approach for choosing a "fast" path.